### DISCUSSION TITLE
Understanding correct usage of `dvf2ddf`

### DISCUSSION THREAD

**Discussion by ArjunNarayanan** [Project-MONAI/MONAI#6807] [Q&A] [answered]
Hi,

My understanding of `dvf2ddf` is that it takes a velocity field and integrates it (via scaling and squaring) to produce a displacement field. The velocity tells me incrementally where a given point will move in the next time step, and the displacement field tells me where a given point will move over several time steps. (Please correct me if I am wrong 😄 )

So if `dvf=1` , and I take 3 steps with `monai.networks.blocks.DVF2DDF` -- I expect `ddf=3` . Here's an example,

```
from monai.networks.blocks import DVF2DDF
import torch

dvf2ddf = DVF2DDF(num_steps=3)
dvf = torch.ones([1,3,10,10,10])

ddf = dvf2ddf(dvf)
ddf_view = ddf.select(2,5)[:,:,3:7,3:7]
```

This gives me

```
tensor([[[[1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.]],
         [[1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.]],
         [[1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.],
          [1., 1., 1., 1.]]]])
```

When I was expecting the interior of `ddf` to have value 3. Am I missing something?

**Comment by ebrahimebrahim** [comment, marked as answer]
Almost the right idea, but there is some confusion.

The `num_steps` argument is not number of time steps but rather number of scaling and squaring steps. To understand scaling and squaring see section 3 of [this paper](https://www-sop.inria.fr/asclepios/Publications/Arsigny/arsigny_miccai06.pdf). In that notation the `num_steps` is $N$.

The amount of "time" is always 1. When you set `num_steps=3` you're saying you want to integrate out from $t=0$ to $t=1$ in $2^N=8$ time steps via $N=3$ squarings. You start with `dvf / 8` as an initial ddf, treating this as an approximation of the ddf that you'd actually get if you integrated your `dvf` from time 0 to time $\frac{1}{8}$. Then you square that ddf (compose it with itself) three times to get to $t=1$.

**Reply by ArjunNarayanan** [reply]
Thanks for your reply, I'll take a look at the paper. Do you know how I can achieve what I am trying to do using `DVF2DDF`?

**Reply by ebrahimebrahim** [reply]
What are you trying to do? If you just want to integrate out to "time = 3" instead of "time = 1" then you can plug in `3*dvf` instead of `dvf`.

**Reply by ArjunNarayanan** [reply]
I'm trying to train a model to perform 3D registration based on [this tutorial](https://github.com/Project-MONAI/tutorials/blob/main/3d_registration/paired_lung_ct.ipynb). Instead of a `ddf` my model predicts a `dvf` so I tried using `DVF2DDF` to convert my `dvf` to a `ddf`. However, this model doesn't work (no sign of the loss going down during training).

I replaced the `Warp` module with my own implementation of grid warping through direct numerical integration (Euler or RK4) and this model does learn a registration field. I'm just wondering why my implementation with `DVF2DDF` does not work.

**Reply by ebrahimebrahim** [reply]
I vaguely remember having successful training with DVF2DDF in the past, but I am not certain. Maybe worth making another post where you share some code and ask why the loss isn't going down, if you're interested in doing that. Could be any number of things I guess :shrug:

**Reply by ArjunNarayanan** [reply]
Yeah that makes sense, I'll do that. Thanks for your help!
